#ЛР №1 Методы интерпретации и визуализации данных
1. Считать набор данных о проишествиях в США (https://www.kaggle.com/datasets/joebeachcapital/homicides?resource=download). 


2. Вывести информацию про:  
    а)первые 5 проишествий;  
    б)все проишествия кроме последних 2-х;  
    в)с 3 по 15 включительно.
    
    
3. Проанализировать на предмет наличия особенностей жертв проишествий, таких как:  
    а) Средний возраст, максимальный и минимальный;  
    б) Узнать количество женщин и мужчин;  
    в) Количество женщин, которым больше 18 лет;  
    г) Количество мужчин, которые были темнокожими. 
    
    
4. Отсортировать всех пассажиров по возрасту по возрастанию и по убыванию и вывести 

In [1]:
import pandas as pd #импортируем библиотеку pandas для обработки dataframe-ов
pd.options.mode.chained_assignment = None 

In [2]:
df = pd.read_csv('homicide-data.csv', encoding="unicode_escape") #Задание №1 - читаем данные
df

uid  reported_date  victim_last victim_first victim_race  \
0      Alb-000001       20100504       GARCIA         JUAN    Hispanic   
1      Alb-000002       20100216      MONTOYA      CAMERON    Hispanic   
2      Alb-000003       20100601  SATTERFIELD      VIVIANA       White   
3      Alb-000004       20100101     MENDIOLA       CARLOS    Hispanic   
4      Alb-000005       20100102         MULA       VIVIAN       White   
...           ...            ...          ...          ...         ...   
52174  Was-001380       20160908     WILLIAMS         EVAN       Black   
52175  Was-001381       20160913        SMITH         DEON       Black   
52176  Was-001382       20161114   WASHINGTON       WILLIE       Black   
52177  Was-001383       20161130       BARNES       MARCUS       Black   
52178  Was-001384       20160901      JACKSON        KEVIN       Black   

      victim_age victim_sex         city state        lat         lon  \
0             78       Male  Albuquerque    NM  35.095788 -106.538555   
1             17       Male  Albuquerque    NM  35.056810 -106.715321   
2             15     Female  Albuquerque    NM  35.086092 -106.695568   
3             32       Male  Albuquerque    NM  35.078493 -106.556094   
4             72     Female  Albuquerque    NM  35.130357 -106.580986   
...          ...        ...          ...   ...        ...         ...   
52174         29       Male   Washington    DC  38.828704  -77.002075   
52175         19       Male   Washington    DC  38.822852  -77.001725   
52176         23       Male   Washington    DC  38.828025  -77.002511   
52177         24       Male   Washington    DC  38.820476  -77.008640   
52178         17       Male   Washington    DC  38.866689  -76.982409   

                 disposition  
0      Closed without arrest  
1           Closed by arrest  
2      Closed without arrest  
3           Closed by arrest  
4      Closed without arrest  
...                      ...  
52174       Closed by arrest  
52175         Open/No arrest  
52176         Open/No arrest  
52177         Open/No arrest  
52178       Closed by arrest  

[52179 rows x 12 columns]

Видно, что в исходном наборе 52179 записей - различных происшествий на территории США, с указанием города, штата, даты проишествия, основной информации о жертве, а также уникальным идентификатором.

Отобразим информацию о первых 5 проишествиях, встречающихся в данном наборе данных с учетом исходной сортировке по UID. Для этого используем функцию head.

In [3]:
df_first_5 = df.head(5)
df_first_5

uid  reported_date  victim_last victim_first victim_race victim_age  \
0  Alb-000001       20100504       GARCIA         JUAN    Hispanic         78   
1  Alb-000002       20100216      MONTOYA      CAMERON    Hispanic         17   
2  Alb-000003       20100601  SATTERFIELD      VIVIANA       White         15   
3  Alb-000004       20100101     MENDIOLA       CARLOS    Hispanic         32   
4  Alb-000005       20100102         MULA       VIVIAN       White         72   

  victim_sex         city state        lat         lon            disposition  
0       Male  Albuquerque    NM  35.095788 -106.538555  Closed without arrest  
1       Male  Albuquerque    NM  35.056810 -106.715321       Closed by arrest  
2     Female  Albuquerque    NM  35.086092 -106.695568  Closed without arrest  
3       Male  Albuquerque    NM  35.078493 -106.556094       Closed by arrest  
4     Female  Albuquerque    NM  35.130357 -106.580986  Closed without arrest

Отобразим информацию о всех проишествиях, кроме последних двух. Для этого используем функцию iloc - она позволяет задать границы выбора df, опираясь на индекс.

In [4]:
df_all_exc_2 = df.iloc[:-2]
df_all_exc_2

uid  reported_date  victim_last victim_first victim_race  \
0      Alb-000001       20100504       GARCIA         JUAN    Hispanic   
1      Alb-000002       20100216      MONTOYA      CAMERON    Hispanic   
2      Alb-000003       20100601  SATTERFIELD      VIVIANA       White   
3      Alb-000004       20100101     MENDIOLA       CARLOS    Hispanic   
4      Alb-000005       20100102         MULA       VIVIAN       White   
...           ...            ...          ...          ...         ...   
52172  Was-001378       20160415         IVEY         PAUL       Black   
52173  Was-001379       20160715       HARRIS       SHAROD       Black   
52174  Was-001380       20160908     WILLIAMS         EVAN       Black   
52175  Was-001381       20160913        SMITH         DEON       Black   
52176  Was-001382       20161114   WASHINGTON       WILLIE       Black   

      victim_age victim_sex         city state        lat         lon  \
0             78       Male  Albuquerque    NM  35.095788 -106.538555   
1             17       Male  Albuquerque    NM  35.056810 -106.715321   
2             15     Female  Albuquerque    NM  35.086092 -106.695568   
3             32       Male  Albuquerque    NM  35.078493 -106.556094   
4             72     Female  Albuquerque    NM  35.130357 -106.580986   
...          ...        ...          ...   ...        ...         ...   
52172         37       Male   Washington    DC  38.826458  -77.003590   
52173         20       Male   Washington    DC  38.827266  -77.001572   
52174         29       Male   Washington    DC  38.828704  -77.002075   
52175         19       Male   Washington    DC  38.822852  -77.001725   
52176         23       Male   Washington    DC  38.828025  -77.002511   

                 disposition  
0      Closed without arrest  
1           Closed by arrest  
2      Closed without arrest  
3           Closed by arrest  
4      Closed without arrest  
...                      ...  
52172       Closed by arrest  
52173         Open/No arrest  
52174       Closed by arrest  
52175         Open/No arrest  
52176         Open/No arrest  

[52177 rows x 12 columns]

Видим что записи с UID Was-001383 и Was-001384 - не отображатся.

Отобразим информацию о проишествиях с 3 по 15 включительно. Для этого используем функцию iloc.

In [5]:
df_3_15 = df.iloc[3:16]
df_3_15

uid  reported_date         victim_last victim_first victim_race  \
3   Alb-000004       20100101            MENDIOLA       CARLOS    Hispanic   
4   Alb-000005       20100102                MULA       VIVIAN       White   
5   Alb-000006       20100126                BOOK    GERALDINE       White   
6   Alb-000007       20100127           MALDONADO        DAVID    Hispanic   
7   Alb-000008       20100127           MALDONADO       CONNIE    Hispanic   
8   Alb-000009       20100130        MARTIN-LEYVA      GUSTAVO       White   
9   Alb-000010       20100210             HERRERA       ISRAEL    Hispanic   
10  Alb-000011       20100212  BARRIUS-CAMPANIONI       HECTOR    Hispanic   
11  Alb-000012       20100218               LUJAN        KEVIN       White   
12  Alb-000013       20100222           COLLAMORE         JOHN    Hispanic   
13  Alb-000014       20100306            CHIQUITO        CORIN       Other   
14  Alb-000015       20100308              TORRES       HECTOR    Hispanic   
15  Alb-000016       20100308                GRAY     STEFANIA       White   

   victim_age victim_sex         city state        lat         lon  \
3          32       Male  Albuquerque    NM  35.078493 -106.556094   
4          72     Female  Albuquerque    NM  35.130357 -106.580986   
5          91     Female  Albuquerque    NM  35.151110 -106.537797   
6          52       Male  Albuquerque    NM  35.111785 -106.712614   
7          52     Female  Albuquerque    NM  35.111785 -106.712614   
8          56       Male  Albuquerque    NM  35.075380 -106.553458   
9          43       Male  Albuquerque    NM  35.065930 -106.572288   
10         20       Male  Albuquerque    NM  35.077375 -106.560569   
11    Unknown       Male  Albuquerque    NM  35.077011 -106.564910   
12         46       Male  Albuquerque    NM  35.064076 -106.608281   
13         16       Male  Albuquerque    NM  35.041491 -106.742147   
14         54       Male  Albuquerque    NM  35.070656 -106.615845   
15         43     Female  Albuquerque    NM  35.070656 -106.615845   

              disposition  
3        Closed by arrest  
4   Closed without arrest  
5          Open/No arrest  
6        Closed by arrest  
7        Closed by arrest  
8          Open/No arrest  
9          Open/No arrest  
10       Closed by arrest  
11  Closed without arrest  
12         Open/No arrest  
13       Closed by arrest  
14       Closed by arrest  
15       Closed by arrest

Определим средний, максимальный и минимальный возраст жертв - атрибут victim_age. Для этого сначала уберем все записи со значением возраста Unknown. По оставшимся записям определим необходимые значения.

In [16]:
df_age = df.query("victim_age != 'Unknown'")
avg_age = df_age['victim_age'].astype(float).mean()
max_age = df_age['victim_age'].astype(float).max()
min_age = df_age['victim_age'].astype(float).min()
data_age = {'Средний возраст':  [avg_age],
        'Максимальный возраст': [max_age],
         'Минимальный возраст': [min_age]
        }

age_table = pd.DataFrame(data_age)

print(age_table)

   Средний возраст  Максимальный возраст  Минимальный возраст
0         31.80122                 102.0                  0.0


Из данных видно, что средний возраст жертвы - 31.8 лет, самой старой жертве - 102 года, самой молодой - было меньше года.

Посчитаем количество мужчин и женщин среди жертв.

In [7]:
man_count = df['victim_sex'].value_counts()['Male']
female_count = df['victim_sex'].value_counts()['Female']
unknw_count = df['victim_sex'].value_counts()['Unknown']
man_percentage = man_count / df['victim_sex'].count() * 100
female_percentage = female_count / df['victim_sex'].count() * 100
unknw_percentage = unknw_count / df['victim_sex'].count() * 100
data_sex = {'Количество': [man_count, female_count, unknw_count],
        '%': [man_percentage, female_percentage, unknw_percentage]
        }

sex_table = pd.DataFrame(data_sex, index=['Man', 'Woman', 'Unknown'])

print(sex_table)


         Количество          %
Man           40739  78.075471
Woman          7209  13.815903
Unknown        4231   8.108626


Из данных видно, что большинство жертв - мужчины, а также что для ~8% жертв - пол скрыт или неизвестен. 

Посчитаем количество женщин, которым больше 18 лет.

In [8]:
df_age['victim_age'] = df_age['victim_age'].astype(float)
df_female = df_age.query("victim_sex == 'Female' & victim_age > 18")
female_18_count = df_female['uid'].count()
print("Количество женщин жертв старше 18 лет -", female_18_count, '\nВ подсчете не учитываются те записи, где пол жертвы скрыт или неизвестен.')

Количество женщин жертв старше 18 лет - 5977 
В подсчете не учитываются те записи, где пол жертвы скрыт или неизвестен.


Подсчитаем количество мужчин-жертв, которые были темнокожими.

In [9]:
df_male = df.query("victim_sex == 'Male'")
black_man_count = df_male['victim_race'].value_counts()['Black']
print("Количество темнокожих мужчин-жертв -", black_man_count, '\nВ подсчете не учитываются те записи, где раса жертвы скрыта или неизвестна.')
black_man_percent_total = black_man_count / df['victim_sex'].count() * 100
black_man_percent_man = black_man_count / man_count * 100
print("Процент темнокожих мужчин-жертв среди всех жертв -", black_man_percent_total, '%', '\nПроцент темнокожих мужчин-жертв среди всех жертв-мужчин - ', black_man_percent_man, '%')


Количество темнокожих мужчин-жертв - 29256 
В подсчете не учитываются те записи, где раса жертвы скрыта или неизвестна.
Процент темнокожих мужчин-жертв среди всех жертв - 56.06853331800149 % 
Процент темнокожих мужчин-жертв среди всех жертв-мужчин -  71.81325020250866 %


Таким образом, больше всего жертв - среди темнокожих мужчин Америки. Почему имеем такое распределение - тема для размышления...

Выведу отсортированные списки по возрасту, без учета тех, чей возраст неизвестен. Если нужно учитовать их - df_age необходимо заменить на df.

In [10]:
df_age.sort_values('victim_age')

uid  reported_date victim_last victim_first victim_race  \
23611  Ind-000423       20110103       SHUAI        ANGEL       Asian   
28530  Lon-000625       20150104    DELACRUZ        ELIZA    Hispanic   
50925  Was-000094       20150814       SMITH       ROBERT       Black   
44526  Pit-000563       20071115     JACKSON      DANIYAH       Black   
22700  Hou-004237       20160425    ROBINSON      BRIELLE       Black   
...           ...            ...         ...          ...         ...   
38808  Oak-000783       20090511        LETT      IVARENE       Black   
4026   Bal-700191       20170721        TATE      WADDELL       Black   
36399   NO-000438       20120223       BRIER         LEON       Black   
524    Atl-000902       20080202       RAMEY          BEN       Black   
50034  Sto-000448       20161007     Unknown      Unknown       Black   

       victim_age victim_sex          city state        lat         lon  \
23611         0.0     Female  Indianapolis    IN  39.912849  -86.077657   
28530         0.0     Female    Long Beach    CA  33.848789 -118.195924   
50925         0.0       Male    Washington    DC  38.936985  -77.025096   
44526         0.0     Female    Pittsburgh    PA  40.460541  -79.987233   
22700         0.0     Female       Houston    TX  29.711726  -95.309567   
...           ...        ...           ...   ...        ...         ...   
38808        97.0     Female       Oakland    CA  37.811889 -122.256068   
4026         97.0       Male     Baltimore    MD  39.314552  -76.593272   
36399        99.0       Male   New Orleans    LA  29.983140  -90.076672   
524         101.0       Male       Atlanta    GA  33.755125  -84.375137   
50034       102.0       Male      Stockton    CA  37.923681 -121.316915   

            disposition  
23611  Closed by arrest  
28530  Closed by arrest  
50925  Closed by arrest  
44526  Closed by arrest  
22700  Closed by arrest  
...                 ...  
38808    Open/No arrest  
4026     Open/No arrest  
36399    Open/No arrest  
524    Closed by arrest  
50034  Closed by arrest  

[49180 rows x 12 columns]

In [11]:
df_age.sort_values('victim_age',ascending=False)

uid  reported_date victim_last victim_first victim_race  \
50034  Sto-000448       20161007     Unknown      Unknown       Black   
524    Atl-000902       20080202       RAMEY          BEN       Black   
36399   NO-000438       20120223       BRIER         LEON       Black   
4026   Bal-700191       20170721        TATE      WADDELL       Black   
38808  Oak-000783       20090511        LETT      IVARENE       Black   
...           ...            ...         ...          ...         ...   
34673  Mil-000955       20170822        REID     SAVAIYAH       Black   
34669  Mil-000951       20110428     SANDERS     ROYALITY       Black   
7146   Cha-000778       20170117        BUSH     MAJESTIC       Black   
23071  Hou-700153       20170803     SOLORIO         ERIK       White   
44497  Pit-000534       20080606      POSTON       MARCIA       Black   

       victim_age victim_sex         city state        lat         lon  \
50034       102.0       Male     Stockton    CA  37.923681 -121.316915   
524         101.0       Male      Atlanta    GA  33.755125  -84.375137   
36399        99.0       Male  New Orleans    LA  29.983140  -90.076672   
4026         97.0       Male    Baltimore    MD  39.314552  -76.593272   
38808        97.0     Female      Oakland    CA  37.811889 -122.256068   
...           ...        ...          ...   ...        ...         ...   
34673         0.0     Female    Milwaukee    wI  43.112005  -88.009101   
34669         0.0       Male    Milwaukee    wI  43.111486  -87.971648   
7146          0.0       Male    Charlotte    NC  35.253339  -80.877047   
23071         0.0       Male      Houston    TX        NaN         NaN   
44497         0.0     Female   Pittsburgh    PA  40.465867  -79.942038   

            disposition  
50034  Closed by arrest  
524    Closed by arrest  
36399    Open/No arrest  
4026     Open/No arrest  
38808    Open/No arrest  
...                 ...  
34673  Closed by arrest  
34669  Closed by arrest  
7146   Closed by arrest  
23071  Closed by arrest  
44497    Open/No arrest  

[49180 rows x 12 columns]

In [14]:
df_age['victim_age'] = df_age['victim_age'].astype(float)
df_female_task = df_age.query("victim_sex == 'Female'")
df_male_task = df_age.query("victim_sex == 'Male'")
print("Средний возраст женщин", df_female_task['victim_age'].mean(),'медианный', df_female_task['victim_age'].median(), 'Средний возраст мужчин', df_male_task['victim_age'].mean(), 'медианный', df_male_task['victim_age'].median())

                         

Средний возраст женщин 34.08997610009841 медианный 31.0 Средний возраст мужчин 31.390431450117443 медианный 28.0
